File: experiments/runs/2026-09-17_00-24-43_claim-office-example-mapping_exact-ptdd-v1.1-refactor-subagent-cc_opus-5-no-thinking-2/src/damaged-item.ts
import type { Item } from "./policy.js";

/**
 * Matching an incident's damage entries to the insured items they hit.
 *
 * This is its own decision, separate from what a damage settles at: it
 * answers "which covered item was this?", not "what does the MHPCO pay?".
 * Each entry consumes a distinct insured item, so an incident naming a type
 * more often than the policy covers it is refused rather than settled -- an
 * entry with no insured item left to consume has nothing to claim against.
 *
 * Which of several alike items an entry consumes is a decision in its own
 * right, and the MHPCO settles it by order: an entry takes the first item of
 * its type the policy still has unconsumed. An incident names a type, never
 * a particular sword, so the MHPCO has no other ground to choose on. The
 * choice is unobservable only while alike items settle alike; two swords at
 * different enchantment levels would make it visible, and it is stated here
 * by name so that revising it means revising this rule rather than
 * discovering it in what the matching happens to do.
 */
export function damagedItems(
  items: readonly Item[],
  itemTypes: readonly string[],
): readonly Item[] {
  const unconsumed = [...items];
  return itemTypes.map((itemType) => {
    const hit = unconsumed.findIndex((item) => item.type === itemType);
    if (hit === -1) {
      throw new Error(
        `This policy does not cover an item of type "${itemType}"`,
      );
    }
    return unconsumed.splice(hit, 1)[0];
  });
}

File: experiments/runs/2026-09-20_10-40-56_claim-office-java-example-mapping_exact-ptdd-v1.1-refactor-subagent-cc_opus-5-no-thinking/src/main/java/DamagedItem.java
import java.util.ArrayList;
import java.util.List;

/**
 * One damage entry of an incident together with the insured item it names.
 *
 * <p>A damage entry names its item by type, the way the report is written;
 * what the clauses need is the insured item itself, with its enchantment and
 * its material. Reading the one against the other is what this record carries:
 * the entry and the item it was matched to, so that no later step has to pair
 * them up again.
 *
 * <p>A policy may cover several items of the same type, so each entry is
 * matched to a separate insured item: the MHPCO will not settle the same item
 * twice on one report, and a report naming more items of a type than the
 * policy covers is refused outright.
 *
 * <p>Changes whenever the MHPCO revises how a report is read against a policy;
 * knows nothing about what is reimbursed for the items it finds.
 */
record DamagedItem(Item insured, Damage damage) {

    /**
     * The damage entries of a report, each paired with the insured item it
     * names, in the order they are reported.
     */
    static List<DamagedItem> of(List<Item> insured, List<Damage> damages) {
        List<Item> unclaimed = new ArrayList<>(insured);
        List<DamagedItem> damaged = new ArrayList<>();
        for (Damage damage : damages) {
            damaged.add(new DamagedItem(claim(unclaimed, damage), damage));
        }
        return damaged;
    }

    private static Item claim(List<Item> unclaimed, Damage damage) {
        for (int index = 0; index < unclaimed.size(); index++) {
            if (unclaimed.get(index).type().equals(damage.itemType())) {
                return unclaimed.remove(index);
            }
        }
        throw new RejectedScenario(
                "the policy does not cover a " + damage.itemType() + " for this damage");
    }
}
